local process = require("@lune/process")
local frktest = require("@packages/frktest")

require("@tests/Buffer.spec")()
require("@tests/Compat.spec")()
require("@tests/SerdeArray.spec")()
require("@tests/SerdeCFrame.spec")()
require("@tests/SerdeColor.spec")()
require("@tests/SerdeColorSequence.spec")()
require("@tests/SerdeEnumItem.spec")()
require("@tests/SerdeFacesAxes.spec")()
require("@tests/SerdeFontContent.spec")()
require("@tests/SerdeNumberSequence.spec")()
require("@tests/SerdeOption.spec")()
require("@tests/SerdePhysicalProperties.spec")()
require("@tests/SerdePrimitives.spec")()
require("@tests/SerdeUDim.spec")()
require("@tests/SerdeRay.spec")()
require("@tests/SerdeUniqueId.spec")()
require("@tests/SerdeVector.spec")()
require("@tests/Artifact.spec")()
require("@tests/SerdeReferent.spec")()
require("@tests/Format.spec")()
require("@tests/ColumnEncoder.spec")()
require("@tests/Instance.spec")()
require("@tests/Cli.spec")()
require("@tests/SchemaEvolution.spec")()
require("@tests/Patch.spec")()
require("@tests/ParallelIntegration.spec")()
require("@tests/ParallelRobloxIntegration.spec")()

local lune_console_reporter = (frktest :: any)._reporters.lune_console_reporter
lune_console_reporter.init()

local success = frktest.run()
process.exit(if success then 0 else 1)
